File: Cindy/Calibration/CalibrationProfile.swift
import Foundation

/// Calibration result for one exercise.
struct ExerciseCalibration: Codable, Equatable, Sendable {
    var source: SignalSource
    var minValue: Float
    var maxValue: Float
    var baseline: Float
    var low: Float
    var high: Float
    var direction: RepDirection
    var repDuration: TimeInterval
    var calibratedAt: Date

    /// Relative to the calibrated baseline; `SignalPipeline` drops the baseline for
    /// signals that do not scale with distance.
    var thresholds: RepThresholds {
        RepThresholds(low: low, high: high, direction: direction, baseline: baseline)
    }

    var range: Float { maxValue - minValue }
}

/// All three exercises; stored as JSON in the Documents directory.
struct CalibrationProfile: Codable, Equatable, Sendable {
    static let currentVersion = 1

    var version: Int = CalibrationProfile.currentVersion
    var createdAt: Date = Date()
    var exercises: [String: ExerciseCalibration] = [:]

    func calibration(for exercise: Exercise) -> ExerciseCalibration? {
        exercises[exercise.rawValue]
    }

    mutating func set(_ calibration: ExerciseCalibration, for exercise: Exercise) {
        exercises[exercise.rawValue] = calibration
    }

    /// Only a profile covering every exercise of the plan can start a workout.
    func isComplete(for plan: WorkoutPlan) -> Bool {
        plan.exercises.allSatisfy { exercises[$0.rawValue] != nil }
    }

    var isComplete: Bool { isComplete(for: .cindy) }

    func missingExercises(for plan: WorkoutPlan) -> [Exercise] {
        plan.exercises.filter { exercises[$0.rawValue] == nil }
    }

    /// Drops calibrations measured on a different signal source than the config uses now
    /// (squats moved from the face to body pose); those exercises need a new calibration.
    func removingOutdated(for config: SignalConfig) -> CalibrationProfile {
        var profile = self
        profile.exercises = exercises.filter { key, calibration in
            guard let exercise = Exercise(rawValue: key) else { return false }
            return calibration.source == config.source(for: exercise)
        }
        return profile
    }
}

/// Loads and saves the calibration profile.
final class CalibrationStore {
    private let store: JSONFileStore<CalibrationProfile>
    private let config: SignalConfig

    init(fileName: String = "calibration.json", config: SignalConfig = .default) {
        store = JSONFileStore(fileName: fileName)
        self.config = config
    }

    init(url: URL, config: SignalConfig = .default) {
        store = JSONFileStore(url: url)
        self.config = config
    }

    /// The stored profile without calibrations for a signal source the config no longer uses.
    func load() -> CalibrationProfile? {
        guard let profile = store.load(), profile.version == CalibrationProfile.currentVersion else { return nil }
        return profile.removingOutdated(for: config)
    }

    func save(_ profile: CalibrationProfile) throws {
        try store.save(profile)
    }

    func delete() {
        store.delete()
    }
}

File: android/core/src/main/kotlin/me/raddatz/cindy/core/calibration/CalibrationProfile.kt
package me.raddatz.cindy.core.calibration

import kotlinx.serialization.Serializable
import me.raddatz.cindy.core.Exercise
import me.raddatz.cindy.core.InstantIso8601Serializer
import me.raddatz.cindy.core.SignalConfig
import me.raddatz.cindy.core.SignalSource
import me.raddatz.cindy.core.WorkoutPlan
import me.raddatz.cindy.core.persistence.JSONFileStore
import me.raddatz.cindy.core.signal.RepDirection
import me.raddatz.cindy.core.signal.RepThresholds
import java.io.File
import java.time.Instant

/** Calibration result for one exercise. [repDuration] in seconds. */
@Serializable
data class ExerciseCalibration(
    val source: SignalSource,
    val minValue: Float,
    val maxValue: Float,
    val baseline: Float,
    val low: Float,
    val high: Float,
    val direction: RepDirection,
    val repDuration: Double,
    @Serializable(with = InstantIso8601Serializer::class)
    val calibratedAt: Instant,
) {
    /**
     * Relative to the calibrated baseline; [me.raddatz.cindy.core.signal.SignalPipeline] drops
     * the baseline for signals that do not scale with distance.
     */
    val thresholds: RepThresholds get() = RepThresholds(low, high, direction, baseline)

    val range: Float get() = maxValue - minValue
}

/** Calibrations of all exercises; stored as JSON. Keys are [Exercise.rawValue]. */
@Serializable
data class CalibrationProfile(
    val version: Int = CURRENT_VERSION,
    @Serializable(with = InstantIso8601Serializer::class)
    val createdAt: Instant = Instant.now(),
    val exercises: Map<String, ExerciseCalibration> = emptyMap(),
) {
    fun calibration(exercise: Exercise): ExerciseCalibration? = exercises[exercise.rawValue]

    /** Swift `set(_:for:)`, returning the changed copy. */
    fun withCalibration(calibration: ExerciseCalibration, exercise: Exercise): CalibrationProfile =
        copy(exercises = exercises + (exercise.rawValue to calibration))

    /** Only a profile covering every exercise of the plan can start a workout. */
    fun isComplete(plan: WorkoutPlan): Boolean = plan.exercises.all { exercises[it.rawValue] != null }

    val isComplete: Boolean get() = isComplete(WorkoutPlan.cindy)

    fun missingExercises(plan: WorkoutPlan): List<Exercise> = plan.exercises.filter { exercises[it.rawValue] == null }

    /**
     * Drops calibrations measured on a different signal source than the config uses now (squats
     * moved from the face to body pose); those exercises need a new calibration.
     */
    fun removingOutdated(config: SignalConfig): CalibrationProfile =
        copy(
            exercises = exercises.filter { (key, calibration) ->
                val exercise = Exercise.fromRawValue(key) ?: return@filter false
                calibration.source == config.source(exercise)
            },
        )

    companion object {
        const val CURRENT_VERSION: Int = 1
    }
}

/** Loads and saves the calibration profile (iOS: `calibration.json` in Documents). */
class CalibrationStore(file: File, private val config: SignalConfig = SignalConfig.default) {
    private val store = JSONFileStore(file, CalibrationProfile.serializer())

    /** The stored profile without calibrations for a signal source the config no longer uses. */
    fun load(): CalibrationProfile? {
        val profile = store.load() ?: return null
        if (profile.version != CalibrationProfile.CURRENT_VERSION) return null
        return profile.removingOutdated(config)
    }

    /** @throws java.io.IOException when the file cannot be written. */
    fun save(profile: CalibrationProfile) {
        store.save(profile)
    }

    fun delete() {
        store.delete()
    }
}
